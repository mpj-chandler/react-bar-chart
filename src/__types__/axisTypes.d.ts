import { SeriesData, NumericDataPoint } from './seriesTypes';
import { Padding } from './styling';
import { Placement } from '../enums/Placement';

export interface AxisRange {
    min: number;
    max: number;
}

export interface AxisProps {
    data: SeriesData[];
    title: string;
    padding: Padding;
    config?: AxisConfig;
}

export interface AxisConfig {
    zeroIntercept: boolean;
    margin: number;
    tickPlacement: Placement;
    tickLength: number;
    titlePos?: {
        xPos: number,
        yPos: number
    }
}

export interface AxisTickProps {
    data: SeriesData[];
    padding: Padding;
    config?: AxisConfig;
}

