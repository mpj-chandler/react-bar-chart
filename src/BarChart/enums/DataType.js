"use strict";
exports.__esModule = true;
var DataType;
(function (DataType) {
    DataType["Named"] = "NAMED";
    DataType["DateIndexed"] = "DATE_INDEXED";
    DataType["Numeric"] = "NUMERIC";
    DataType["NonNullNumeric"] = "NON_NULL_NUMERIC";
})(DataType || (DataType = {}));
;
exports["default"] = DataType;
