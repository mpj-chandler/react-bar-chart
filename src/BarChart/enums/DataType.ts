enum DataType {
    Named = 'NAMED',
    DateIndexed = 'DATE_INDEXED',
    Numeric = 'NUMERIC',
    NonNullNumeric = 'NON_NULL_NUMERIC',
};

export default DataType;
