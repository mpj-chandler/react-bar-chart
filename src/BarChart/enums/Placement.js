"use strict";
exports.__esModule = true;
var Placement;
(function (Placement) {
    Placement[Placement["Aligned"] = 0] = "Aligned";
    Placement[Placement["Bucket"] = 1] = "Bucket";
})(Placement || (Placement = {}));
exports["default"] = Placement;
