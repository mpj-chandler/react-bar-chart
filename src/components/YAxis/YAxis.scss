.YAxis {
  grid-area: plot;
  z-index: 1;

  &__Svg {
    width: 100%;
    height: 100%;
  }

  &__Ticks {
    width: 100%;
    height: 100%;
  }

  &__TickLabels {
    font-size: calc(8px + 0.5vh);
  }
}
