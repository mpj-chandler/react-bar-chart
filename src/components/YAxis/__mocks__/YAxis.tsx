import React from 'react';

const YAxis: React.FC<any> = (props: any) => <div {...props}>Mock YAxis {props.children}</div>

export default YAxis;