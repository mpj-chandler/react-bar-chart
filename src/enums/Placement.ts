enum Placement {
    Aligned,
    Bucket,
}

export default Placement;
