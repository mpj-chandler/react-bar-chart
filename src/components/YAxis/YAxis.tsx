import React from 'react';
import styles from './YAxis.scss';
import { getYAxisRange } from '../../utils/axisUtils/getYAxisRange/getYAxisRange';
import { getXAxisRange } from '../../utils/axisUtils/getXAxisRange/getXAxisRange';
import generateAxisLabels from '../../utils/axisUtils/generateAxisLabels/generateAxisLabels';
import Axis from '../../enums/Axis';
import AxisTickLabel from '../AxisTickLabel/AxisTickLabel';
import { AxisProps, AxisTickProps, AxisRange } from '../../__types__/axisTypes';
import DataType from '../../enums/DataType';
import AxisTitleLabel from '../AxisTitleLabel/AxisTitleLabel';

function getYAxisXPos(props: AxisProps): number {
    const range = getXAxisRange(props.data, props.config);

    return (100 * (range.min / (range.max - range.min)) + props.padding.left);
}

function renderTicks(props: AxisTickProps, x2: number): JSX.Element {
    const tickLabels: {
        dataType: DataType,
        labels: Array<{index: number, label: string }>,
    } = generateAxisLabels(props.data, Axis.YAxis);

    const range: AxisRange = getYAxisRange(props.data, props.config);
    const interval = range.max - range.min;
    const tickLength = props.config ? props.config.tickLength : 5;

    return (
        <g className={styles.YAxis__Ticks}>
            {tickLabels.labels.map((item: { index: number, label: string }) => {
                const y1 = (((100 - (props.padding.top + props.padding.bottom))
                 * (1 - (Number(item.label) - range.min) / interval)))
                 + props.padding.top;

                return (
                    <g key={item.label}>
                        <line
                            x1={`${x2 - tickLength}%`}
                            y1={`${y1}%`}
                            x2={`${x2}%`}
                            y2={`${y1}%`}
                            stroke={'black'}
                        />
                        <AxisTickLabel
                            key={`tickText-${item.label}`}
                            xPos={x2 - tickLength}
                            yPos={y1}
                            label={item.label}
                            axis={Axis.YAxis}
                        />
                    </g>
                );
            })}
        </g>
    );
};

const YAxis: React.FC<AxisProps> = (props) => {

    const yAxisXPos = getYAxisXPos(props);
    const titlePos =
      props.config && props.config.titlePos ? props.config.titlePos : undefined;

    return (
        <div className={styles.YAxis}>
            <svg className={styles.YAxis__Svg}>
                <line x1={`${yAxisXPos}%`} y1={`${props.padding.top}%`} x2={`${yAxisXPos}%`} y2={`${100 - props.padding.bottom}%`} stroke={'black'} fill={'transparent'} strokeWidth={1}/>
                {renderTicks({data: props.data, padding: props.padding, config: props.config}, yAxisXPos)}
                <AxisTitleLabel titlePos={titlePos} label={props.title} axis={Axis.YAxis}/>
            </svg>
        </div>
    );
};

export default YAxis;
