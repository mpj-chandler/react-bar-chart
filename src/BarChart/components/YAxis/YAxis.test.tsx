import React from 'react';
import TestRenderer from 'react-test-renderer';
import YAxis from './YAxis';
import Placement from '../../enums/Placement';
import { AxisProps, AxisConfig } from '../../__types__/axisTypes';
import DataType from '../../enums/DataType';


jest.mock('../AxisTickLabel/AxisTickLabel');

describe('YAxis', () => {
    const props: AxisProps = {
        title: 'anything',
        data: [
            {
                seriesName: 'A',
                type: {
                    x: DataType.NonNullNumeric,
                    y: DataType.NonNullNumeric,
                },
                points: [
                    {
                        x: 0,
                        y: 10,
                    },
                    {
                        x: 1,
                        y: 20,
                    },
                    {
                        x: 2,
                        y: 30,
                    }
                ]
            }
        ],
        padding: {
            top: 10,
            bottom: 10,
            left: 10,
            right: 10,
        }
    };

    const config: AxisConfig = {
        zeroIntercept: true,
        margin: 20,
        tickPlacement: Placement.Aligned,
        tickLength: 10,
    }

    it('it renders consistently', () => {
        const component = TestRenderer.create(<YAxis {...props}/>);

        expect(component.toJSON()).toMatchSnapshot();
    });

    it('it renders consistently with an explicit config', () => {
        const component = TestRenderer.create(<YAxis {...props} config={config}/>);

        expect(component.toJSON()).toMatchSnapshot();
    });
})