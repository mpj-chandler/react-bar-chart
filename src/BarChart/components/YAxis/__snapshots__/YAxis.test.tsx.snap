// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YAxis it renders consistently 1`] = `
<div
  className="YAxis"
>
  <svg
    className="YAxis__Svg"
  >
    <line
      fill="transparent"
      stroke="black"
      strokeWidth={1}
      x1="10%"
      x2="10%"
      y1="10%"
      y2="90%"
    />
    <g
      className="YAxis__Ticks"
    >
      <g>
        <line
          stroke="black"
          x1="5%"
          x2="10%"
          y1="130%"
          y2="130%"
        />
        <div
          axis={1}
          label="0"
          xPos={5}
          yPos={130}
        >
          MockAxisTickLabel 
        </div>
      </g>
      <g>
        <line
          stroke="black"
          x1="5%"
          x2="10%"
          y1="90%"
          y2="90%"
        />
        <div
          axis={1}
          label="10"
          xPos={5}
          yPos={90}
        >
          MockAxisTickLabel 
        </div>
      </g>
      <g>
        <line
          stroke="black"
          x1="5%"
          x2="10%"
          y1="50%"
          y2="50%"
        />
        <div
          axis={1}
          label="20"
          xPos={5}
          yPos={50}
        >
          MockAxisTickLabel 
        </div>
      </g>
      <g>
        <line
          stroke="black"
          x1="5%"
          x2="10%"
          y1="10%"
          y2="10%"
        />
        <div
          axis={1}
          label="30"
          xPos={5}
          yPos={10}
        >
          MockAxisTickLabel 
        </div>
      </g>
    </g>
  </svg>
</div>
`;

exports[`YAxis it renders consistently with an explicit config 1`] = `
<div
  className="YAxis"
>
  <svg
    className="YAxis__Svg"
  >
    <line
      fill="transparent"
      stroke="black"
      strokeWidth={1}
      x1="10%"
      x2="10%"
      y1="10%"
      y2="90%"
    />
    <g
      className="YAxis__Ticks"
    >
      <g>
        <line
          stroke="black"
          x1="0%"
          x2="10%"
          y1="90%"
          y2="90%"
        />
        <div
          axis={1}
          label="0"
          xPos={0}
          yPos={90}
        >
          MockAxisTickLabel 
        </div>
      </g>
      <g>
        <line
          stroke="black"
          x1="0%"
          x2="10%"
          y1="67.77777777777777%"
          y2="67.77777777777777%"
        />
        <div
          axis={1}
          label="10"
          xPos={0}
          yPos={67.77777777777777}
        >
          MockAxisTickLabel 
        </div>
      </g>
      <g>
        <line
          stroke="black"
          x1="0%"
          x2="10%"
          y1="45.55555555555556%"
          y2="45.55555555555556%"
        />
        <div
          axis={1}
          label="20"
          xPos={0}
          yPos={45.55555555555556}
        >
          MockAxisTickLabel 
        </div>
      </g>
      <g>
        <line
          stroke="black"
          x1="0%"
          x2="10%"
          y1="23.33333333333333%"
          y2="23.33333333333333%"
        />
        <div
          axis={1}
          label="30"
          xPos={0}
          yPos={23.33333333333333}
        >
          MockAxisTickLabel 
        </div>
      </g>
    </g>
  </svg>
</div>
`;
